### PR TITLE
Setting up data for analysis!!

### PR DESCRIPTION
**Pull request by cansavvy**
# Description
First of many pull requests to translate the steps of the workflow here https://github.com/FredHutch/GI_mapping/blob/main/workflow/scripts/pgRNA_counts_QC.Rmd to a functionalized and reproducible version.

### CODE REVIEW

File: vignettes/getting-started.Rmd [hunk 1]
```diff
@@ -17,14 +17,64 @@ knitr::opts_chunk$set(
...
+example_pg_metadata <- example_data %>%
+  dplyr::select(c("id", "seq_1", "seq_2"))
+
+# sample metadata is the information that describes 
+example_sample_metadata <- data.frame(
+  day = as.factor(c("Day00", "Day05", "Day22", "Day22", "Day22")), 
+  rep = as.factor(c("RepA", "RepA", "RepA", "RepB", "RepC"))
+)
+```
+
```
Comment (cansavvy) [inline vignettes/getting-started.Rmd:61]:
```suggestion
Tidbit
```

File: R/00-setup_data.R [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    metadata = list(pg_metadata = NULL,
+                    sample_metadata = NULL)
+  )
+
+  class(new_data) <- c("list", "gimap_dataset")
+
+  if (is.null(counts)) stop("counts cannot be NULL")
+  if (!is.matrix(counts)) stop("counts can only be in the form of a matrix")
+
+  # If they don't give sample metadata, then we will make up a row id
```
Comment (kweav) [inline R/00-setup_data.R:36]: This is the equivalent of a column ID, right?

File: vignettes/getting-started.Rmd [hunk 3]
```diff
@@ -17,13 +17,65 @@ knitr::opts_chunk$set(
...
+```{r}
+example_counts <- example_data %>%
+  dplyr::select(c("Day00_RepA", "Day05_RepA", "Day22_RepA", "Day22_RepB", "Day22_RepC")) %>%
+  as.matrix()
+```
+
+The next two datasets are metadata that describe the dimensions of the count data. 
+- These both need to be data frames. 
+- The sizes of these metadata must correspond to the dimensions of the counts data. 
+- The first column of the pg_metadata must be a unique id
```
Comment (kweav) [inline vignettes/getting-started.Rmd:48]: Do we need to set up a check for this in the `setup_data()` function?

File: R/00-setup_data.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R:36]: The same concept, but it won't be stored as a column.

File: R/00-setup_data.R [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  if (is.null(sample_metadata)) sample_metadata <- 1:nrow(counts)
+  if (!is.data.frame(sample_metadata)) stop("metadata can only be in the form of a data.frame")
+  if (nrow(sample_metadata) != ncol(counts)) stop("the number of rows in the sample metadata is not equal to the number of columns in the counts" )
+
+  new_data$metadata$sample_metadata <- sample_metadata
+
+  if (!is.null(pg_metadata)) {
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R:43]: I suppose that makes sense? Will do.

File: R/00-setup_data.R [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    new_data$metadata$pg_metadata <- pg_metadata
+  }
+  # Store the raw counts
+  new_data$raw_counts <- counts
+
+  # Calculate the counts per sample
+  new_data$counts_per_sample <- apply(counts, 2, sum)
+
+  # Calculate the coverage per pgRNA
+  new_data$coverage <- new_data$counts_per_sample/nrow(counts)
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R:56]: I feel like no need to round so soon -- might as well keep the significant digits until later.

[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R:56]: We should ask the Berger lab about the significance of the 31700. On a quick look I thought it was the row count but you're right its 2000 off. I think I asked them about this but I'm not seeing it on our conversations.

[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R:56]: @danieljgroso Do you know the significance of the [31700 in this line in the original code](https://github.com/FredHutch/GI_mapping/blob/main/workflow/scripts/pgRNA_counts_QC.Rmd) ?

File: R/00-setup_data.R [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  # Transform the data
+  new_data$transformed_data$count_norm <- -log10((counts + 1)/sum(counts))
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R:59]: Ah. Good catch. Thanks!

Comment (cansavvy) [inline R/00-setup_data.R:59]:
```suggestion
  new_data$transformed_data$count_norm <- apply(counts, 2, function(x) -log10((x+1)/sum(x)))
```

File: R/utils.R [hunk 7]
```diff
@@ -15,3 +17,53 @@ example_data <- function() {
...
+}
+
+print_kbl <- function(tbl) {
+  kbl(tbl) %>%
+    kable_styling(full_width = FALSE,
+                  position = "left",
+                  bootstrap_options = c("striped", "hover", "responsive"))
+}
+
+save_tbl <- function(tbl){
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/utils.R:41]: These functions I directly copied from https://github.com/FredHutch/GI_mapping/blob/1b3c4cd48193f7570518d1a2327ea55f5bed637f/workflow/scripts/shared_functions_and_variables.R but I haven't necessarily updated them yet.

File: vignettes/getting-started.Rmd (hunk 3 above)
Comment (cansavvy) [reply vignettes/getting-started.Rmd:48]: Yes! Good call. I'll add that.

Comment (cansavvy) [reply vignettes/getting-started.Rmd:48]: Added in https://github.com/FredHutch/gimap/pull/12/commits/655a7fb2ace7775ede0b91167c8a1bedfb2fb54d

File: R/00-setup_data.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R]: Fixed in https://github.com/FredHutch/gimap/pull/12/commits/add7a3fc0ff0908003c2b9e91479303a2048c7dd

File: R/00-setup_data.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (cansavvy) [reply R/00-setup_data.R]: Added in https://github.com/FredHutch/gimap/pull/12/commits/f79b4485728835b758f410e2b85112eab99cc4c2

**Review by kweav — APPROVED** [review]

**Comment by cansavvy** [comment]
@kweav All fixed! (sort of) but the run_qc() stuff needs help on the next PR.

**Comment by cansavvy** [comment]
> @kweav All fixed! (sort of) but the run_qc() stuff needs help on the next PR.

Feel free to change any of the code in the utils.R file to make run_qc() function work more seamlessly!

**Pull request merged**